set authentication configuration in Pulsar client (#45)

atezs82 opened the pull request:
This PR attempts to finish the almost-complete authentication for the Pulsar connector. Briefly the following were done:
- `CachedPulsarClient` and `PulsarMetadataReader` sets authentication parameters in the Pulsar clients.
- Auth options in `PulsarOptions` were converted to camel-case, so they can be picked properly by eg. the `AdminUtils.buildAdmin` method.
- A new configuration module type called `pulsar.admin` is added. If given, this configuration is used when creating the admin interface in `PulsarMetaDataReader`. If not, the client configuration will be used.
- Added documentation about the `pulsar.admin` configuration module and some code examples on how to configure authentication when connecting.

During local testing the modified code could authenticate towards to a Pulsar cluster and receive messages from there successfully.

Please let me know what do you think about this draft approach. I will add proper unit testing around this if this can be a good way forward.

Thanks in advance!

sijie commented:
@syhily @jianyun8023 Can you review this pull request?

syhily commented:
Thanks for your contribution, I will review it later tonight.

syhily commented on src/main/scala/org/apache/spark/sql/pulsar/PulsarProvider.scala:
```diff
@@ -368,6 +378,23 @@ private[pulsar] object PulsarProvider extends Logging {
     }
   }
 
+  private def getAdminParams(parameters: Map[String, String]): Map[String, String] = {
```
This method is copied multiple times. Could you just merge these three `getProducerParams`, `getReaderParams`, `getAdminParams` method into one common method.

atezs82 replied:
That is a good idea indeed, fixed this.

atezs82 replied:
That's a good idea indeed, fixed this.

syhily commented on src/main/scala/org/apache/spark/sql/pulsar/PulsarProvider.scala:
```diff
@@ -562,18 +589,20 @@ private[pulsar] object PulsarProvider extends Logging {
     var clientParams = getClientParams(parameters)
     clientParams += (SERVICE_URL_OPTION_KEY -> serviceUrl)
     val readerParams = getReaderParams(parameters)
+    val adminParams = if (hasAdminParams(parameters)) getAdminParams(parameters) else clientParams
```
I don't think this is a good idea. `hasAdminParams` just call the `getAdminParams` and filter. This would cause the real getAdminParams be called twice if we give some pulsar admin config.

syhily replied:
Use `Option(getAdminParams(parameters)).filter(_.nonEmpty).getOrElse(clientParams)` would be better.

atezs82 replied:
Ok, fixed this.

syhily commented on src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala:
```diff
@@ -478,4 +479,26 @@ private[pulsar] case class PulsarMetadataReader(
         }
     }
   }
+  private def createPulsarClient(
```
This method share the same logic with `CachedPulsarClient`, why not using this class instead?

atezs82 replied:
Fixed this (now the class is using a `CachedPulsarClient` instead).

syhily commented on src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala:
```diff
@@ -392,7 +393,7 @@ private[pulsar] case class PulsarMetadataReader(
           assert (time > 0, s"time less than 0: $time")
           if (client == null) {
-            client = PulsarClient.builder().serviceUrl(serviceUrl).build()
+            client = createPulsarClient(serviceUrl, clientConf)
```
https://github.com/streamnative/pulsar-spark/blob/87b1cb8a7c62cf5100a91a5ba568ad17997483c2/src/main/scala/org/apache/spark/sql/pulsar/PulsarMetadataReader.scala#L462 This line should also be modified to use `createPulsarClient`

atezs82 replied:
This had been removed, since we will be using a `CachedPulsarClient` for this as well.

syhily commented on src/main/scala/org/apache/spark/sql/pulsar/CachedPulsarClient.scala:
```diff
@@ -62,7 +65,11 @@ private[pulsar] object CachedPulsarClient extends Logging {
+  def createPulsarClient(
+                          pulsarConf: ju.Map[String, Object],
+                          pulsarClientBuilder: ClientBuilder = PulsarClient.builder()): Client = {
+    def hasConfig(key: String) = pulsarConf.get(key) != null
```
Method `hasConfig` is unused. Plz check if you need it, otherwise just remove it.

atezs82 replied:
Removed it.

syhily commented on src/main/scala/org/apache/spark/sql/pulsar/CachedPulsarClient.scala:
```diff
@@ -62,7 +65,11 @@ private[pulsar] object CachedPulsarClient extends Logging {
       .build[Seq[(String, Object)], Client](cacheLoader)
 
-  private def createPulsarClient(pulsarConf: ju.Map[String, Object]): Client = {
+  def createPulsarClient(
```
This method is exposed, why? I guess that you would like to use it in `PulsarMetadataReader.scala`.

atezs82 replied:
This had been exposed so that the test of the class could access it. My original idea was not to fire up an authentication-enabled Pulsar cluster for this purpose (just for unit testing this class), I wanted check the authentication parameters are set properly in the client builder only.

syhily commented on src/main/scala/org/apache/spark/sql/pulsar/PulsarConfigUpdater.scala:
```diff
@@ -37,6 +38,10 @@ private[pulsar] case class PulsarConfigUpdater(
   def set(key: String, value: Object, map: ju.Map[String, Object]): this.type = {
     if (blacklistedKeys.contains(key)) {
       logInfo(s"$module: Skip $key")
+    } else if (keysToHideInLog.contains(key)) {
```
Since the `keysToHideInLog` is just hide the log. How about drop this `else if` statement and merge them into one.

```scala
  def set(key: String,
          value: Object,
          map: ju.Map[String, Object]): this.type = {
    if (blacklistedKeys.contains(key)) {
      logInfo(s"$module: Skip $key")
    } else {
      map.put(key, value)
      logInfo(
        s"$module: Set $key to ${printConfigValue(key, Option(value))}," +
          s" earlier value: ${printConfigValue(key, pulsarParams.get(key))}"
      )
    }
    this
  }


  private val HIDE_COMPLETELY_LIMIT = 6
  private val SHOW_FRACTION_OF_HIDDEN_VALUE = 1.0 / 3.0

  private def printConfigValue(key: String,
                               maybeVal: Option[Object]): String = {
    val value = maybeVal.map(_.toString).getOrElse("")
    if (keysToHideInLog.contains(key)) {
      if (value.length < HIDE_COMPLETELY_LIMIT) {
        return "...<completely hidden>..."
      } else {
        return s"${value.take((value.length * SHOW_FRACTION_OF_HIDDEN_VALUE).toInt)}..."
      }
    }

    value
  }
```

atezs82 replied:
I have changed the code accordingly.

syhily commented on src/main/scala/org/apache/spark/sql/pulsar/PulsarConfigUpdater.scala:
```diff
@@ -76,4 +81,14 @@ private[pulsar] case class PulsarConfigUpdater(
     map
   }
 
+  private val HIDE_COMPLETELY_LIMIT = 6
```
This is not a static field, so just use camel case by following scala style.

atezs82 replied:
Done.

syhily commented on src/test/scala/org/apache/spark/sql/pulsar/CachedPulsarClientBuilderSuite.scala:
```diff
@@ -0,0 +1,53 @@
+import org.scalatest.mockito.MockitoSugar.mock
+
+class CachedPulsarClientBuilderSuite extends SparkFunSuite {
+  test("Should pass on authentication related parameters when authentication plugin is provided") {
```
Could you add an integration test by using `TestContainer`.

atezs82 replied:
Currently the test suite is working locally due to multiple problems:
- "Too many open files" is getting thrown from some tests
- "Failed to remove cursor" warnings are thrown from PulsarMetadataReader

Please give me some time to fix these issues. If anything obvious comes up that might be an improvement to the documentation (so that others do not run into these problems) I'll add them to the documentation as well.

atezs82 replied:
Currently the test suite is working locally due to multiple problems:

- "Too many open files" is getting thrown from some tests (setting `docker-ulimits` to some very high value does not help) 
- "Failed to remove cursor" warnings are thrown from PulsarMetadataReader (albeit this might be normal, since the test otherwise succeeds).

Please give me some time to fix these issues. If you already know these and can share some insights on them, that would be most appreciated. I am not entirely sure whether these are caused by my modifications or is there something else misconfigured in my environment, but there are problems coming even when I just clone a clean master and do a `mvn clean install`. If anything comes up that might be an improvement to the documentation (so that others do not run into these problems) I'll add them to there as well.

There had been an other problem as well, a clean clone of the master gives back the following:
```
com.github.dockerjava.api.exception.ConflictException: {"message":"Conflict. The container name \"/standalone-standalone\" is already in use by container \"...long container hash...\". You have to remove (or rename) that container to be able to reuse that name."}
```
We can fix this by appending a sufficiently long random suffix to the name of the container. That made me pass this error but hit the others above.

syhily reviewed (changes requested):
Thank you for your contribute. Your code is perfect, but a little style issues need to be improve.

atezs82 commented:
@syhily Thanks for the review, I'll fix the issues and push the changes soon.

atezs82 reviewed (commented):
@syhily I have fixed most of the issues, but need more time to do the integration test you asked for. See reasoning above, and thanks again for the review!

atezs82 commented:
@syhily Fixed all issues, and pushed up the code (just waiting for the code check to finish). Did the following things briefly regarding your integration test comment:

- Added two simple integration tests for `CachedPulsarClient` to test whether it can:
    - Set authentication properties in the client properly (using a mock Client builder).
    - Work with a locally running Pulsar cluster using authentication.
- Changed `PulsarContinuousReader`, `PulsarMicroBatchReader` and `PulsarSource` to close down their `PulsarMetadataReader` only once. Consecutive calls to `stop` in the metadata reader caused exceptions being thrown during tests.
- Set a random container name when starting test containers in `PulsarTest` (see that issue above in my comment). Also, clean up the service during shutdown (apparently this did not help in solving the `Too many files` problem outlined above).
- Modified slightly the documentation (added a hint about executing individual test suites).

Please let me know what do you think and thanks in advance for reviewing this!

syhily commented:
You PR is awesome, I don't find any extra issues.

syhily commented:
I will merge this PR after the Github Action is added. The jenkins is not available anymore.

atezs82 commented:
@syhily Thanks for the comments and the review!
